Import component for specific platform with an alias

**Do you want to request a *feature* or report a *bug*?**
Report a bug

**What is the current behavior?**
Using react-native, I have a component for each platform `MyComponent.android.js` and `MyComponent.ios.js`. I have `package.json` in my `src` folder like this `{name: "src"}` to be able to import my components like this: `src/components/MyComponent`.
When testing a component that import MyComponent like this: `import MyComponent from 'src/components/MyComponent'`, I get this error: `Cannot find module 'components/MyComponent' from 'index.js'`.
It seems like Jest is not able to resolve the import for specific platform when there is an alias.
I also tried with `babel-plugin-module-resolver` instead of a `package.json` but no luck.

If I change the import and put a relative path instead of an absolute one, it is working.

**What is the expected behavior?**
Jest should be able to resolve component import for specific platform when using an alias for path.

**Please provide your exact Jest configuration and mention your Jest, node, yarn/npm version and operating system.**
RN 0.43.2
Jest 19.0.2

Configuration:
```
"jest": {
    "preset": "react-native",
    "setupFiles": [
      "./.jestsetup.js"
    ],
    "transformIgnorePatterns": [
      "node_modules/(?!(jest-)?react-native|react-navigation)"
    ],
    "snapshotSerializers": [
      "<rootDir>/node_modules/enzyme-to-json/serializer"
    ]
  }
```

```
// .jestsetup.js

// Make Enzyme functions available in all test files without importing
import React from "react-native";
import { shallow, render, mount } from "enzyme";
global.shallow = shallow;
global.render = render;
global.mount = mount;

// mock fetch
global.fetch = jest.fn().mockImplementation(() => Promise.resolve());

// Fail tests on any warning
console.error = message => {
  throw new Error(message);
};

jest.mock("Linking", () => {
  return {
    addEventListener: jest.fn(),
    removeEventListener: jest.fn(),
    openURL: jest.fn(),
    canOpenURL: jest.fn(),
    getInitialURL: jest.fn()
  };
});

React.NativeModules.RNBranch = {};
```

--- a/integration_tests/resolve/__tests__/resolve-test.js
+++ b/integration_tests/resolve/__tests__/resolve-test.js
@@ -38,6 +38,11 @@ test('should resolve filename.<platform>.js', () => {
   expect(platform.extension).toBe('android.js');
 });
 
+test('should resolve filename.<platform>.js from haste package', () => {
+  expect(testRequire('custom-resolve/test1')).not.toThrow();
+  expect(platform.extension).toBe('android.js');
+});
+
 test('should resolve filename.native.js', () => {
   expect(testRequire('../test2')).not.toThrow();
   expect(platform.extension).toBe('native.js');

--- a/integration_tests/resolve/package.json
+++ b/integration_tests/resolve/package.json
@@ -1,4 +1,5 @@
 {
+  "name": "custom-resolve",
   "jest": {
     "haste": {
       "platforms": ["native"],

--- a/packages/jest-resolve/src/index.js
+++ b/packages/jest-resolve/src/index.js
@@ -138,8 +138,8 @@ class Resolver {
     const skipResolution =
       options && options.skipNodeResolution && !moduleName.includes(path.sep);
 
-    if (!skipResolution) {
-      module = Resolver.findNodeModule(moduleName, {
+    const resolveNodeModule = name => {
+      return Resolver.findNodeModule(name, {
         basedir: dirname,
         browser: this._options.browser,
         extensions,
@@ -147,6 +147,10 @@ class Resolver {
         paths,
         resolver: this._options.resolver,
       });
+    };
+
+    if (!skipResolution) {
+      module = resolveNodeModule(moduleName);
 
       if (module) {
         return (this._moduleNameCache[key] = module);
@@ -156,12 +160,17 @@ class Resolver {
     // 3. Resolve "haste packages" which are `package.json` files outside of
     // `node_modules` folders anywhere in the file system.
     const parts = moduleName.split('/');
-    module = this.getPackage(parts.shift());
-    if (module) {
+    const hastePackage = this.getPackage(parts.shift());
+    if (hastePackage) {
       try {
-        return (this._moduleNameCache[key] = require.resolve(
-          path.join.apply(path, [path.dirname(module)].concat(parts)),
-        ));
+        const module = path.join.apply(
+          path,
+          [path.dirname(hastePackage)].concat(parts),
+        );
+        // try resolving with custom resolver first to support extensions,
+        // then fallback to require.resolve
+        return (this._moduleNameCache[key] =
+          resolveNodeModule(module) || require.resolve(module));
       } catch (ignoredError) {}
     }
 

